File: leetcode/exercises/2.2.1.easy.valid_parenthesis.js
// -----------------------
// Created    : 
// Last Edited: 
// Topics     : 
// Big O      :
// Problem Id : 20
// -----------------------

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    map = {
        "}":"{",
        ")":"(",
        "]":"[",
    };

    stack = [];
    for (const i of s) {
        if (i in map) {
            if (
                (stack.length == 0) ||
                (stack[stack.length - 1] != map[i])
            ) {
                return false;
            }
            stack.pop();
        } else {
            stack.push(i);
        }
    }

    if (stack.length == 0) { return true } 
    return false

};

console.log(isValid("(){}}{"))

File: leetcode/exercises/2.2.1.easy.valid_parenthesis.py
# -----------------------
# Created     : 17/11/2025
# Last Edited : 17/11/2025 
# Topics      : 
# Big O       :
# Problem Id  : 20
# Source      :
# Notes       :
# -----------------------

class Solution:
    def isValid(self, s: str) -> bool:
        map = {
                "}": "{",
                ")": "(",
                "]": "[",
        }
        stack = []
        for i in s:
            if i in map:
                if len(stack) == 0 or stack[-1] != map.get(i, False):
                    return False
                stack.pop()
            else:
                stack.append(i)

        return True if len(stack) == 0 else False 
        

s = Solution()
print(s.isValid("(){}}{"))

  
